client/discovery/discovery.go

```diff
@@ -0,0 +1,133 @@
+package discovery
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/nomad/client/config"
+	"github.com/hashicorp/nomad/nomad/structs"
+)
+
+// Builtins is a set of discovery layer implementations for various
+// providers. Each provider must implement the Provider interface.
+var Builtins = []Factory{
+	NewConsulDiscovery,
+}
+
+// Factory is a function interface to create new discovery providers.
+type Factory func(Context) (Provider, error)
+
+// Provider is a generic interface which can be used to implement
+// service discovery back-ends in Nomad.
+type Provider interface {
+	// Name returns the type of the service discovery subsystem. This
+	// is used to identify the system in log messages and usually
+	// returns a static string value.
+	Name() string
+
+	// Enabled determines if the discovery layer has been enabled.
+	Enabled() bool
+
+	// DiscoverName returns the name used to register the task into service
+	// discovery. The return value is used in log messages and is passed to
+	// the Register and Deregister functions.
+	DiscoverName(parts []string) string
```

Comment:
I would prefer if `parts []string` was replaced with a struct that could be filled by the TaskRunner.